src/cmd/services/m3coordinator/server/m3msg/handler.go

```diff
@@ -0,0 +1,155 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3msg
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/m3db/m3metrics/encoding/msgpack"
+	"github.com/m3db/m3msg/consumer"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/log"
+
+	"github.com/uber-go/tally"
+)
+
+// Options for the ingest handler.
+type Options struct {
+	InstrumentOptions         instrument.Options
+	WriteFn                   WriteFn
+	AggregatedIteratorOptions msgpack.AggregatedIteratorOptions
+}
+
+type handlerMetrics struct {
+	messageReadError                tally.Counter
+	metricAccepted                  tally.Counter
+	droppedMetricMsgpackDecodeError tally.Counter
+	droppedMetricDecodeMalformed    tally.Counter
+}
+
+func newHandlerMetrics(scope tally.Scope) handlerMetrics {
+	messageScope := scope.SubScope("metric")
+	return handlerMetrics{
+		messageReadError: scope.Counter("message-read-error"),
+		metricAccepted:   messageScope.Counter("accepted"),
+		droppedMetricMsgpackDecodeError: messageScope.Tagged(map[string]string{
+			"reason": "msgpack-decode-error",
+		}).Counter("dropped"),
+		droppedMetricDecodeMalformed: messageScope.Tagged(map[string]string{
+			"reason": "decode-malformed",
+		}).Counter("dropped"),
+	}
+}
+
+type handler struct {
+	writeFn      WriteFn
+	iteratorOpts msgpack.AggregatedIteratorOptions
+	logger       log.Logger
+	m            handlerMetrics
+}
+
+func newHandler(opts Options) (*handler, error) {
+	return &handler{
+		writeFn:      opts.WriteFn,
+		iteratorOpts: opts.AggregatedIteratorOptions,
+		logger:       opts.InstrumentOptions.Logger(),
+		m:            newHandlerMetrics(opts.InstrumentOptions.MetricsScope()),
+	}, nil
+}
+
+func (h *handler) Handle(c consumer.Consumer) {
+	h.newPerConsumerHandler().handle(c)
+}
+
+func (h *handler) newPerConsumerHandler() *perConsumerHandler {
+	return &perConsumerHandler{
+		writeFn: h.writeFn,
+		logger:  h.logger,
+		m:       h.m,
+		it:      msgpack.NewAggregatedIterator(nil, h.iteratorOpts),
+		r:       bytes.NewReader(nil),
+	}
+}
+
+type perConsumerHandler struct {
+	// Per server variables, shared across consumers/connections.
+	writeFn WriteFn
+	logger  log.Logger
+	m       handlerMetrics
+
+	// Per consumer/connection variables.
+	it msgpack.AggregatedIterator
+	r  *bytes.Reader
+}
+
+func (h *perConsumerHandler) handle(c consumer.Consumer) {
+	var (
+		msgErr error
+		msg    consumer.Message
+	)
+	for {
+		msg, msgErr = c.Message()
+		if msgErr != nil {
+			break
+		}
+
+		h.processMessage(msg)
+	}
+	if msgErr != nil && msgErr != io.EOF {
```

Comment:
how will the loop get restarted

Reply:
the consumer is long lived with a tcp connection, c.Message() is a blocking call and the loop continues to call it to keep decoding messages from the connection